### PR TITLE
[#127150963] Adds the opportunity to sign up/in through FB

### PR DESCRIPTION
**Pull request by Leonidas777**
https://www.pivotaltracker.com/n/projects/1724033/stories/127150963

### CODE REVIEW

File: config/initializers/000_omnyauth.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Rails.application.config.middleware.use OmniAuth::Builder do
```
Comment (dmitirystrukov) [inline config/initializers/000_omnyauth.rb]: omniauth имя файла?

Comment (Leonidas777) [inline config/initializers/000_omnyauth.rb]: No, that's the special class that's used to work with omniauth.

File: db/migrate/20160730103458_create_social_profiles.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class CreateSocialProfiles < ActiveRecord::Migration
+  def change
+    create_table :social_profiles do |t|
+      t.integer :user_id
```
Comment (dmitirystrukov) [inline db/migrate/20160730103458_create_social_profiles.rb]: null: false ?

File: lib/modules/omniauthable.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def self.find_or_create_with_oauth(oauth_data)
+      find_with_oauth(oauth_data) || create_with_oauth(oauth_data)
+    end
+
+    def self.find_with_oauth(oauth_data)
+      find_through_authorization(oauth_data.provider, oauth_data.uid)
+    end
+
+    def self.find_through_authorization(service_name, uid)
+      social_profile = SocialProfile.where(service_name: service_name, uid: uid).first
```
Comment (dmitirystrukov) [inline lib/modules/omniauthable.rb]: find_by

File: lib/modules/omniauthable.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    def self.parse_name(oauth_data)
+      if oauth_data.provider == 'facebook'
+        oauth_data.extra.raw_info.name.split(' ')
+      end
+    end
+
+    def register_social_profile(service_name, uid)
+      social_profile = SocialProfile.where(service_name: service_name, uid: uid).first_or_create
```
Comment (dmitirystrukov) [inline lib/modules/omniauthable.rb]: find_by

File: lib/modules/omniauthable.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      if social_profile.user_id.present?
+        return false if social_profile.user_id != id
```
Comment (dmitirystrukov) [inline lib/modules/omniauthable.rb]: надо эту проверку в сексуальный метод поместить social_profile.user_id != id

File: config/application.rb [hunk 6]
```diff
@@ -13,3 +13,5 @@ class Application < Rails::Application
     # -- all .rb files in that directory are automatically loaded.
   end
 end
+
+require Rails.root.join 'lib', 'modules', 'omniauthable'
```
Comment (dmitirystrukov) [inline config/application.rb]: магия какая то обьясни пожалуйста

File: app/views/layouts/partials/_header.html.haml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+%header
+  %h3 The ketties of the Day
+  - if user_signed_in?
+    %span= "Hello, #{current_user.first_name} #{current_user.last_name}"
```
Comment (dmitirystrukov) [inline app/views/layouts/partials/_header.html.haml]: модель должна иметь метод full_name

File: app/controllers/omniauth_callbacks_controller.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    sign_in_with_oauth_data(oauth_data) unless user_signed_in?
+
+    current_user.register_social_profile(oauth_data.provider, oauth_data.uid)
+
+    redirect_to root_path
+  end
+
+  def sign_in_with_oauth_data(data)
+    user = User.find_or_create_with_oauth(data)
+    sign_in :user, user
```
Comment (dmitirystrukov) [inline app/controllers/omniauth_callbacks_controller.rb]: sign_in :user, User.find_or_create_with_oauth(data)
Почему не сделать так?

File: config/application.rb (hunk 6 above)
Comment (Leonidas777) [inline config/application.rb]: We just include the Omniauthable module while running our application. That module is in charge to handle what we get from FB. It has logic for OmniauthCallback.

File: lib/modules/omniauthable.rb (hunk 3 above)
Comment (Leonidas777) [inline lib/modules/omniauthable.rb]: Explain me, please why I think so.

File: app/controllers/omniauth_callbacks_controller.rb (hunk 8 above)
Comment (Leonidas777) [inline app/controllers/omniauth_callbacks_controller.rb]: It doesn't look readable.

File: lib/modules/omniauthable.rb (hunk 5 above)
Comment (Leonidas777) [inline lib/modules/omniauthable.rb]: There is no need to do that. In this case it's acceptable.

File: lib/modules/omniauthable.rb (hunk 3 above)
Comment (dmitirystrukov) [inline lib/modules/omniauthable.rb]: такая конструкция короче будет, не нужен first

File: lib/modules/omniauthable.rb (hunk 5 above)
Comment (dmitirystrukov) [inline lib/modules/omniauthable.rb]: @Leonidas777 почему?

File: app/controllers/omniauth_callbacks_controller.rb (hunk 8 above)
Comment (dmitirystrukov) [inline app/controllers/omniauth_callbacks_controller.rb]: @Leonidas777 ну с этим в принципе согласен, оставь как есть

File: lib/modules/omniauthable.rb (hunk 3 above)
Comment (Leonidas777) [inline lib/modules/omniauthable.rb]: Ok, I agree now.

File: lib/modules/omniauthable.rb (hunk 5 above)
Comment (Leonidas777) [inline lib/modules/omniauthable.rb]: You will have to create an extra method either in the SocialProfile model or current module and that method will take a social profile as a parameter. Doesn't it look too hard for such short condition? I think this small solution won't make it easier.

**Pull request merged**